Fix #18366: E2E Flake : Node outline save button is not visible

hiramatsuyuusuke opened the pull request:
I fixed `story-node-editor.component.html`.
E2E Test Error on `topicAndStoryEditorFileUploadFeatures` was fixed.

## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #[18366].
2. This PR does the following: [This PR fix `story-node-editor.component.html`. Because to fix the E2E Test Error on `topicAndStoryEditorFileUploadFeatures` ]
3. (For bug-fixing PRs only) The original bug occurred because: [Node outline save button is not visible. This PR removed the code to make the button disappear.]

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] My PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).


## Proof that changes are correct

![pass](https://github.com/oppia/oppia/assets/140389350/4d5c6a13-8b7e-4598-acff-a7ac60ccf159)

[manually_10_24.webm](https://github.com/oppia/oppia/assets/140389350/0dab08db-1a8c-4056-83b1-5a453ceae0d1)

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be
sure that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network
In my local environment, I needed to change the time of the timeout in etoe test, but I haven't changed the timeout time because etoe on github doesn't need it. And the same goes for cookie consent.
<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone
It was stated on the wiki that we do not currently do mobile etoe testing.
<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language
I did not change the UI shape, so I think it is ok.
<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

hiramatsuyuusuke committed:
Fix E2E Flake : Node outline save button is not visible

I fixed `story-node-editor.component.html`.
E2E Test Error on `topicAndStoryEditorFileUploadFeatures` was fixed.

hiramatsuyuusuke commented:
@Lawful2002 PTAL

seanlip reviewed (changes requested):
@hiramatsuyuusuke Could you please show a video of the behaviour here?

In particular, is the save button container shown even when there are no edits to save? That would be weird behaviour so I just wanted to double-check (typically, that button should probably only be shown if we're in "edit" mode).

hiramatsuyuusuke commented:
@seanlip OK. Is it okay with etoe videos? And I would be happy if you could give me some tips on how to make a video.

seanlip commented:
Hi @hiramatsuyuusuke, I think it is better if you do a manual screencast. You should be able to start the server using the instructions [here](https://github.com/oppia/oppia/wiki/Installing-Oppia) and set up the necessary data using the instructions [here](https://github.com/oppia/oppia/wiki/LaCE-onboarding-guide). Can you try that?

For local screencasts, it depends on your machine, but see if you have something like Quicktime Player (Mac/Windows) or SimpleScreenRecorder (Linux).

hiramatsuyuusuke commented:
Thank you for your advice @seanlip. I first need to learn how to get to `Topics and Skill Dashboard`. I think it will take some time, so please tell me the deadline for submitting the video.

seanlip commented:
@hiramatsuyuusuke Please do it as soon as you're able, and feel free to ask on GitHub Discussions if you need help. Thanks!

hiramatsuyuusuke commented:
Hello, I got the `etoe test` video earlier, so I show it here. Then I will try a manual screencast, but please let me know if you don't need it.
In the video, the save button is initially light color and  I think the save button is not effective. When characters are entered into the text, the color changes to dark and  I think the save button is effective. 

https://github.com/oppia/oppia/assets/140389350/2b1dbdcf-f6d0-49ac-aa9c-975cee992d3b

seanlip commented:
@hiramatsuyuusuke Thanks! I think the issue with the current UI is that it's not clear what the "Cancel" button does in this case. Shouldn't Save/Cancel only show when the user is in "editing" mode? I think the issue here is with the underlying code rather than with the e2e test.

If you look at the exploration editor page (accessible from /creator-dashboard when creating a new exploration) there is a similar UI for the exploration content that shows a readonly view, but when you click on it to edit, it changes to an editable view, and when you click save/cancel, it changes to a readonly view again. Maybe that would be a better option?

Lawful2002 commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.html:
```diff
@@ -136,7 +136,7 @@
                                    [localValue]="editableOutline"
                                    (localValueChange)="updateLocalEditableOutline($event)">
               </schema-based-editor>
-              <div class="save-button-container" *ngIf="chapterOutlineButtonsAreShown">
```
Same concern that Sean rasied, removing this bit of code might not be the expected behaviour here.

hiramatsuyuusuke replied:
Hello @Lawful2002. I understand. I'm currently working on the fix. Thanks!

Lawful2002 reviewed (changes requested):
Thanks @hiramatsuyuusuke, same concern that Sean rasied.

hiramatsuyuusuke committed:
Fix behavior of chapter outline and related e2e.

hiramatsuyuusuke commented:
Hello, I fixed the behavior as shown below. And this fixs passed `etoe` test on `topicAndStoryEditorFileUploadFeatures`. I would like a review of this fix.
[manually.webm](https://github.com/oppia/oppia/assets/140389350/8e2694fb-f602-46f3-ac87-bb06c80cbd8b)

hiramatsuyuusuke committed:
Fix lint mistake and e2e mistake.

I fix lint mistakes. Then I fix a e2e mistake in "story-node-editor.component.ts" line 550.

hiramatsuyuusuke committed:
I fixed inline style in "story-node-editor.component.html".

hiramatsuyuusuke committed:
I fixed alphabetical order and space and from <label> to </label> in html.

hiramatsuyuusuke committed:
I fixed <label> indent.

hiramatsuyuusuke committed:
I added cancel button action in function of `changeNodeOutline`.

hiramatsuyuusuke committed:
I fixed trailing sapces again.

hiramatsuyuusuke committed:
Remove `toggleChapterOutlineButtons()` in `story-node-editor.component.ts`.

hiramatsuyuusuke committed:
Remove `it('should toggle chapter outline buttons', () => {});` in '~.spec.ts'

hiramatsuyuusuke committed:
I added 'should click chapter outline save button' and 'cancel' to `~spec.ts`

hiramatsuyuusuke committed:
Fix  `it('should click chapter outline save button'` and `cancel` in `~spec.ts`.

hiramatsuyuusuke committed:
Fix it('should click chapter outline save button' and `cancel`, second.

hiramatsuyuusuke committed:
Fix `onSaveButtonClicked()` in `story-node-editor.component.ts`.

hiramatsuyuusuke committed:
Fix `onSaveButtonClicked()` in `story-node-editor.component.ts`, second.

hiramatsuyuusuke committed:
Fix `onSaveButtonClicked()` in `story-node-editor.component.ts`, 3rd.

hiramatsuyuusuke committed:
remove pencil icon

hiramatsuyuusuke committed:
Added `action.click()` to `changeNodeOutline` in`StoryEditorPage.js`.

hiramatsuyuusuke committed:
Fix `Outline Finalized` in `story-node-editor.component.html`.

hiramatsuyuusuke committed:
Fix ngOnInit() in `story-node-editor.component.ts`. And revert pencil icon.

hiramatsuyuusuke committed:
I go back to the 1st fix to verify the karuma error.

hiramatsuyuusuke committed:
I added `onSaveButtonClicked()` to verify the karuma error.

hiramatsuyuusuke committed:
I fixed 'onSaveButtonClicked()' to verify the karma error.

hiramatsuyuusuke committed:
I fixed 'onSaveButtonClicked()' to verify the karma error, second.

hiramatsuyuusuke committed:
I fixed `onSaveButtonClicked()` to verify the karma error, 3rd.

hiramatsuyuusuke committed:
Fix to verify the karma error, 4th.

hiramatsuyuusuke committed:
Fix to verify the karma error, 5th.

hiramatsuyuusuke committed:
I go back `Fix `ngOnInit()`.  Added  `updateLocalEditableOutlines` to `~spec.ts`.

hiramatsuyuusuke committed:
Fix `Outline Finalized` in html. Fix `changeNodeOutline` in `StoryEditorPage.js`.

hiramatsuyuusuke committed:
Fix `changeNodeOutline` in `StoryEditorPage.js`, again.

hiramatsuyuusuke committed:
Fix ` it('should update updateLocalEditableOutlines'`.

hiramatsuyuusuke committed:
Fix it('should update `updateLocalEditableOutlines`, second. Fix StoryEditorPage.js

hiramatsuyuusuke commented:
I will make an explanation of the code and check the code, again. Because the code was changed to pass tests here.

hiramatsuyuusuke committed:
Update story-node-editor.component.ts

hiramatsuyuusuke committed:
I go back to the previous passed code.

hiramatsuyuusuke committed:
the original code

hiramatsuyuusuke commented:
I fixed the code to pass CI tests as shown below. Please review this fix. @
seanlip, @seanlip,@Lawful2002, @DubeySandeep, or someone PTAL.
![explanation_of_code](https://github.com/oppia/oppia/assets/140389350/3e1f3eaf-4eb4-427c-8996-ed8e13d89155)

seanlip commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.html:
```diff
@@ -117,31 +117,47 @@
           </div>
-          <div class="chapter-card-inputs oppia-mobile-collapsible-card-content" *ngIf="chapterOutlineIsShown">
+          <br>
+          <div class="outline-Finalized-container" *ngIf="oldOutline">
```
Please follow existing conventions for CSS classes -- it should be kebab-case with individual words starting with lowercase letters.

Also, why do you need ngIf="oldOutline" here -- what happens if there's no oldOutline?

hiramatsuyuusuke replied:
I fix kebab-case and remove `ngIf="oldOutline"`. Then I replace the second PR video with the video that show this fix behavior on 10/8.
![fix_html1](https://github.com/oppia/oppia/assets/140389350/3088377b-7bbd-45de-a44f-49ee807e97ea)

seanlip commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.html:
```diff
@@ -117,31 +117,47 @@
-            <div tabindex="-1" (blur)="toggleChapterOutlineButtons()" (click)="toggleChapterOutlineButtons()">
+          </div>
+          <div *ngIf="!chapterOutlineIsShown">
+            <div class="outline-content-container" *ngIf="!oldOutlineToReadOnly">
```
The name `oldOutlineToReadOnly` is not clear, and I also don't understand what its type is. 

Maybe you mean !outlineEditViewIsShown or similar? (Conceptually, there should only be one outline item, and it's either editable or readonly -- oldOutline refers to a previous version of that outline.)

hiramatsuyuusuke replied:
I replaced `oldOutlineToReadOnly` with `outlineEditViewIsShown`.
![fix_html_ts_1](https://github.com/oppia/oppia/assets/140389350/fd96cb07-0a44-46fb-b594-98da2aa6f3ac)

seanlip commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.html:
```diff
@@ -117,31 +117,47 @@
+            <div class="outline-content-container" *ngIf="!oldOutlineToReadOnly">
+              <input class="chapter-outline-editor-content e2e-test-node-outline-editor-content e2e-test-rte"
+                     readonly
+                     (click)="chapterOutlineIsShown=true"
```
Please replace this with a makeChapterOutlineEditable() function or similar, we try to avoid doing direct operations in the html files.

hiramatsuyuusuke replied:
I replace `chapterOutlineIsShown=true` with `makeChapterOutlineEditable()`.
![fix_html_ts_2](https://github.com/oppia/oppia/assets/140389350/91ca3644-d633-4dc1-9c0f-d14f92f85d52)

seanlip commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.html:
```diff
@@ -117,31 +117,47 @@
+                     readonly
+                     (click)="chapterOutlineIsShown=true"
+                     placeholder="Open chapter outline editor.">
+              <label class="fas fa-pencil-alt oppia-icon" (click)="chapterOutlineIsShown=true"></label>
```
Ditto.

hiramatsuyuusuke replied:
Ditto.

seanlip commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.html:
```diff
@@ -117,31 +117,47 @@
+                     placeholder="Open chapter outline editor.">
+              <label class="fas fa-pencil-alt oppia-icon" (click)="chapterOutlineIsShown=true"></label>
+            </div>
+            <div class="outline-content-container" *ngIf="oldOutlineToReadOnly" >
```
See above re naming, please also remove the space before ">"

hiramatsuyuusuke replied:
I replace `oldOutlineToReadOnly` with `outlineEditViewIsShown`. Then I remove the space.
![fix_html_2](https://github.com/oppia/oppia/assets/140389350/b5dc7dda-9c9d-4fca-8e84-3ff08025abe4)

seanlip commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.html:
```diff
@@ -117,31 +117,47 @@
                         class="btn btn-success save-button e2e-test-node-outline-save-button"
                         [disabled]="!isOutlineModified(editableOutline)"
-                        (click)="updateOutline(editableOutline)">
+                        (click)="updateOutline(editableOutline); onSaveButtonClicked();">
```
Can updateOutline(editableOutline) be included in onSaveButtonClicked() (and removed from here), or will this break something else?

seanlip replied:
@hiramatsuyuusuke did you see this comment?

seanlip replied:
@hiramatsuyuusuke Just flagging this again. Could you please respond? Thanks.

seanlip commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.ts:
```diff
@@ -496,15 +497,20 @@ export class StoryNodeEditorComponent implements OnInit, OnDestroy {
-  toggleChapterOutlineButtons(): void {
-    this.chapterOutlineButtonsAreShown = !this.chapterOutlineButtonsAreShown;
+  onSaveButtonClicked(): void {
+    this.oldOutlineToReadOnly = $(this.oldOutline).text();
```
Please try to avoid use of jQuery here. 

Also, why do you need to do this? What does "oldOutlineToReadOnly" represent, exactly?

hiramatsuyuusuke replied:
Hi @seanlip. I use jQuery to remove html tag  `<p></p>`. I'm stuck on how to remove the tag. Could I  have some advice?

seanlip commented on core/tests/webdriverio_utils/StoryEditorPage.js:
```diff
@@ -448,6 +449,8 @@ var StoryEditorPage = function() {
   };
 
   this.changeNodeOutline = async function(richTextInstructions) {
+    await browser.execute('window.scrollTo(0,500);');
```
Is this necessary? We don't do it in our other tests so why do we need it here?

hiramatsuyuusuke replied:
I remove this at `I fixed UI layout on 10/9, 3rd`.

seanlip replied:
This seems to be still in the code, though? (It shows up in "Files Changed".)

seanlip reviewed (changes requested):
Hi @hiramatsuyuusuke, thanks for your PR! I left a few comments.

Also, could you please show a comprehensive video demonstrating the current behaviour after your changes? Feel free to replace the video in the original PR description with the latest version.

Thanks!

hiramatsuyuusuke commented:
Hi @seanlip, I replace first video with the one of 10/6.

hiramatsuyuusuke committed:
First half of the fix of the review.

seanlip commented:
Hi @hiramatsuyuusuke, thanks for the video! Some notes:

- When you open the editor, there's a huge margin between the top of it and the chapter outline. Please get rid of that margin and make sure that the left edge of the rich-text editor aligns with the left edge of "Chapter Outline".
- When you open the editor, the Save and Cancel buttons should show up immediately, not after a while. Just disable the "Save" button if no change to the existing content has been made.
- It is fine for the "Outline finalized" checkbox to always show. Just make sure that its left edge coincides vertically with the left edge of "Chapter Outline", and that there's enough vertical space between it and the readonly / editable text.

Could you please make these changes and the other remaining fixes, so that I can review again? The above should hopefully simplify some of the logic in your code, too.

Finally, just a note, it's fine to not add images of the code you've edited (though you can if you want). In general I can figure everything out by looking at the file changes in the "Files changed" tab. Just wanted to let you know -- thanks!

hiramatsuyuusuke commented:
I fix UI layout on 10/9.
[ui_layout_10_9.webm](https://github.com/oppia/oppia/assets/140389350/a873f6e8-38bc-44aa-8035-d19c51bc58d2)

hiramatsuyuusuke committed:
I fixed UI layout on 10/9.

hiramatsuyuusuke committed:
I fixed UI layout on 10/9, second.

hiramatsuyuusuke committed:
I fixed UI layout on 10/9, 3rd.

hiramatsuyuusuke committed:
I go back to `First half of the fix of the review`.

hiramatsuyuusuke committed:
I fixed UI layout, again.

seanlip commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.html:
```diff
@@ -117,31 +117,45 @@
                                    [localValue]="editableOutline"
-                                   (localValueChange)="updateLocalEditableOutline($event)">
+                                   (localValueChange)="updateLocalEditableOutline($event);"
+                                   (click)="chapterOutlineButtonsAreShown=true; ">
```
Do you still need the "click" action here? I don't think so, right?

hiramatsuyuusuke replied:
I'm fixing again, because there are CI tests problems. And the other problems are same.

hiramatsuyuusuke replied:
I fixed this. Then the fixed code passed CI tests at `remove chapterOutlineButtonsAreShown` on 10/10.

seanlip commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.ts:
```diff
@@ -496,15 +497,24 @@ export class StoryNodeEditorComponent implements OnInit, OnDestroy {
-  toggleChapterOutlineButtons(): void {
-    this.chapterOutlineButtonsAreShown = !this.chapterOutlineButtonsAreShown;
+  onSaveButtonClicked(): void {
+    this.outlineEditViewIsShown = $(this.oldOutline).text();
```
Why do you need this line at all? Also, I think that outlineEditViewIsShown should be a boolean and not a string since the name indicates true/false...

Additionally I am confused by the type of this.oldOutline -- it seems like it's supposed to be a string so I'm not sure why you're interpreting it using jQuery. Are you trying to check whether it's empty? If so, would `!this,oldOutline` suffice?

hiramatsuyuusuke replied:
I used jQuery because  `outlineEditViewIsShown` is used to display the saved text as shown below.
Also html tag is ramained without jQuery.
![remain_html_tag](https://github.com/oppia/oppia/assets/140389350/95f49efe-e1e4-449f-bb07-b163d2edd10a)

seanlip reviewed (changes requested):
Hi @hiramatsuyuusuke -- took another look, and the UI looks really good! Left a few code comments -- if you could wrap those up, let's try to get this in ASAP since the CI error is still appearing for developers.

But, overall, it is looking really good. Thanks a lot for fixing this and making sure that the user experience is a nice one!

hiramatsuyuusuke committed:
I fixed UI layout, again, second.

hiramatsuyuusuke committed:
I fixed UI layout, again, 3rd.

hiramatsuyuusuke committed:
I fixed UI layout, again, 4th.

hiramatsuyuusuke committed:
Update story-node-editor.component.html

hiramatsuyuusuke committed:
I fixed UI layout, again, 6th.

hiramatsuyuusuke committed:
remove `chapterOutlineButtonsAreShown`.

hiramatsuyuusuke committed:
remove `await browser.execute('window.scrollTo(0,500);');`.

hiramatsuyuusuke committed:
remove `await browser.execute('window.scrollTo(0,500);');`, second.

hiramatsuyuusuke committed:
remove await `browser.execute('window.scrollTo(0,500);');`, 3rd.

hiramatsuyuusuke committed:
remove `await browser.execute('window.scrollTo(0,500);');`, 4th.

hiramatsuyuusuke committed:
move `updateOutline()` into `onSaveButtonClicked()`.

hiramatsuyuusuke committed:
move `updateOutline()` into `onSaveButtonClicked()`, second.

hiramatsuyuusuke committed:
move `updateOutline()` into `onSaveButtonClicked()`, 3rd.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), 4th.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), 5th.

hiramatsuyuusuke committed:
Update story-node-editor.component.ts

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), 7th.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), 8th.

hiramatsuyuusuke committed:
I go back to `move updateOutline() into onSaveButtonClicked()`.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), again.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), again, second.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), again, 3rd.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), again, 4th.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), again, 5th.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), again, 6th.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), again, 7th.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), again, 8th.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), again, 9th.

hiramatsuyuusuke committed:
move updateOutline() into onSaveButtonClicked(), again, 10th.

hiramatsuyuusuke commented:
I could remove jQuery `$().text()` and the code passed CI test on 10/17. Then HTML tags were displayed as shown below.
[10_17.webm](https://github.com/oppia/oppia/assets/140389350/f46db22b-6c27-4495-8382-9e0e06f887ca)

hiramatsuyuusuke committed:
remove ` $().text();`.

hiramatsuyuusuke committed:
I go back to `10th`.

hiramatsuyuusuke committed:
I go back, further.

hiramatsuyuusuke committed:
fix $().text();.

hiramatsuyuusuke committed:
fix $().text();, second.

hiramatsuyuusuke committed:
fix $().text();, 3rd.

hiramatsuyuusuke commented:
Hello, I'm stuck on how to display the text of the saved content in HTML mode. I think I should use something like   `OUTLINE_SCHEMA = {
    type: 'html',` that is described by `<schema-based-editor>` tag. However I don't know which tags are valid and which properties are valid. Could someone please give me some advice?

seanlip commented:
@hiramatsuyuusuke Sorry, could you please give a bit more detail about the issue? What do you mean by "how to display the text of the saved content in HTML mode" (what exactly are you trying to do)?

Thanks!

hiramatsuyuusuke commented:
The rich text editor string is getted along with the html tag. Therefore, currently the html tag is displayed along with the string as shown fig.1 or the video of `10_17.webm`. However I think that the html tags should be enabled like fig.2. So, I think `type: 'html';` is useful. For example,  `type: 'html';` exist in `interface HtmlSchema {}` of `schema-default-value.service.ts`. Or `type: 'html';` exist in `story-node-editor.component.ts`. However now I don't know how to use  `type: 'html';`, correctly.
![fig 1,2](https://github.com/oppia/oppia/assets/140389350/a50ae43c-c9b9-49e3-b57d-709d11786116)

seanlip commented:
Ah, I see. We used to use angular-html-bind for this but I think that got
changed. Can you look at the exploration player page for learners to see
how they display the lesson content? You'll want to do a similar thing
here, I think.



On Thu, Oct 19, 2023, 11:56 AM Yuusuke Hiramatsu ***@***.***>
wrote:

> The rich text editor string is getted along with the html tag. Therefore,
> currently the html tag is displayed along with the string as shown fig.1 or
> the video of 10_17.webm. However I think that the html tags should be
> enabled like fig.2. So, I think type: 'html'; is useful. For example, type:
> 'html'; exist in interface HtmlSchema {} of
> schema-default-value.service.ts. Or type: 'html'; exist in
> story-node-editor.component.ts. However now I don't know how to use type:
> 'html'; .
> [image: fig 1,2]
> <https://user-images.githubusercontent.com/140389350/276570208-a50ae43c-c9b9-49e3-b57d-709d11786116.jpg>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/18899#issuecomment-1770567702>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQV5SSRZTX4QSIU3EXWVO3YAEBNNAVCNFSM6AAAAAA46WQOLSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONZQGU3DONZQGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

seanlip commented:
(I think it's the conversation skin or tutor card component)

On Thu, Oct 19, 2023, 2:37 PM Sean Lip ***@***.***> wrote:

> Ah, I see. We used to use angular-html-bind for this but I think that got
> changed. Can you look at the exploration player page for learners to see
> how they display the lesson content? You'll want to do a similar thing
> here, I think.
>
>
>
> On Thu, Oct 19, 2023, 11:56 AM Yuusuke Hiramatsu ***@***.***>
> wrote:
>
>> The rich text editor string is getted along with the html tag. Therefore,
>> currently the html tag is displayed along with the string as shown fig.1 or
>> the video of 10_17.webm. However I think that the html tags should be
>> enabled like fig.2. So, I think type: 'html'; is useful. For example, type:
>> 'html'; exist in interface HtmlSchema {} of
>> schema-default-value.service.ts. Or type: 'html'; exist in
>> story-node-editor.component.ts. However now I don't know how to use type:
>> 'html'; .
>> [image: fig 1,2]
>> <https://user-images.githubusercontent.com/140389350/276570208-a50ae43c-c9b9-49e3-b57d-709d11786116.jpg>
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/oppia/oppia/pull/18899#issuecomment-1770567702>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACQV5SSRZTX4QSIU3EXWVO3YAEBNNAVCNFSM6AAAAAA46WQOLSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONZQGU3DONZQGI>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

hiramatsuyuusuke committed:
fix $().text();, 4th.

hiramatsuyuusuke committed:
fix $().text();, 5th.

hiramatsuyuusuke committed:
fix $().text();, 6th.

hiramatsuyuusuke committed:
fix $().text();, 7th.

hiramatsuyuusuke committed:
I go back to 3rd.

hiramatsuyuusuke committed:
fix $().text();, again.

hiramatsuyuusuke committed:
fix $().text();, again, second.

hiramatsuyuusuke committed:
fix $().text();, again, 3rd.

hiramatsuyuusuke committed:
fix $().text();, again, 4th.

hiramatsuyuusuke committed:
fix $().text();, again, 5th.

hiramatsuyuusuke committed:
fix $().text();, again, 6th.

hiramatsuyuusuke committed:
Update StoryEditorPage.js

hiramatsuyuusuke committed:
Update StoryEditorPage.js

hiramatsuyuusuke committed:
fix $().text();, again, 9th.

hiramatsuyuusuke commented:
I fixed the code to pass CI tests. Please review this fix. @seanlip,@Lawful2002, @DubeySandeep, or someone PTAL.

hiramatsuyuusuke commented:
I update the fixed code behavior video.
[10_24.webm](https://github.com/oppia/oppia/assets/140389350/8c51623d-1cd5-49cb-897d-21a6d3983ba1)

DubeySandeep commented on core/tests/webdriverio_utils/StoryEditorPage.js:
```diff
@@ -473,7 +473,8 @@ var StoryEditorPage = function() {
     var outlineEditorRteContentText = (
       await nodeOutlineEditorRteContent[0].getText());
-    expect(outlineEditorRteContentText).toEqual(nodeOutline);
+    var nodeOutlineText = await nodeOutline.replace(/<.+?>/g, '');
```
Considering `nodeOutline` is a function param, why can't we pass the required value and test it? Instead of changing it here before assert.

Note: This method will succesfully pass the test for both `'<p>First outline</p>'`, `'<random>First outline</random>'`.

hiramatsuyuusuke replied:
Thank you for the review, @DubeySandeep. I'm working on it.

DubeySandeep reviewed (changes requested):
@hiramatsuyuusuke Thanks for the PR, I've reviewed the two e2e test files and left one comment, PTAL!

hiramatsuyuusuke committed:
fix `StoryEditorPage.js` by using `[value]` in `story-node-editor.component.html`

hiramatsuyuusuke committed:
fix `<p></p>`

hiramatsuyuusuke committed:
fix lint.

hiramatsuyuusuke committed:
I go back to 10/24.

hiramatsuyuusuke committed:
fix  `expect('<p>' + outlineEditorRteContentText + '</p>')`

hiramatsuyuusuke commented:
I fixed the change of `nodeOutline`.  Please review this fix. @seanlip,@Lawful2002, @DubeySandeep, or someone PTAL.

U8NWXD reviewed (approved):
E2E changes LGTM, but we'll need @DubeySandeep to remove their blocking review before the PR can merge

DubeySandeep commented on core/tests/webdriverio_utils/StoryEditorPage.js:
```diff
@@ -473,7 +473,7 @@ var StoryEditorPage = function() {
     var outlineEditorRteContentText = (
       await nodeOutlineEditorRteContent[0].getText());
-    expect(outlineEditorRteContentText).toEqual(nodeOutline);
+    expect('<p>' + outlineEditorRteContentText + '</p>').toEqual(nodeOutline);
```
Avoid hardcoding the `expect` block, I would suggest matching the text part instead to be safe.

```
this.expectNodeOutlineTextToMatch = async function(nodeOutline) {
    var nodeOutlineEditorRteContent = (
      await nodeOutlineEditorRteContentSelector());
    var outlineEditorRteContentText = (
      await nodeOutlineEditorRteContent[0].getText());
    expect( outlineEditorRteContentText).toEqual(nodeOutline);
    ```
    
    Should we do this?

DubeySandeep replied:
Avoid hardcoding the `expect` block, I would suggest matching the text part instead to be safe.

```
this.expectNodeOutlineTextToMatch = async function(nodeOutline) {
    var nodeOutlineEditorRteContent = (
      await nodeOutlineEditorRteContentSelector());
    var outlineEditorRteContentText = (
      await nodeOutlineEditorRteContent[0].getText());
    expect( outlineEditorRteContentText).toEqual(nodeOutline);
```    

Should we do this?

DubeySandeep commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.ts:
```diff
@@ -534,7 +541,7 @@ export class StoryNodeEditorComponent implements OnInit, OnDestroy {
   ngOnInit(): void {
     this.pageTitleService.setNavbarTitleForMobileView('Chapter Editor');
-    this.chapterOutlineIsShown = !this.windowDimensionsService.isWindowNarrow();
+    this.chapterOutlineIsShown = false;
```
Is this the reason for the button's disappearance? If yes, have we looked into why window was narrow in CI?

Also, do we need to handle all other parts (line 545 to 550) similarly?

DubeySandeep replied:
The initial value of `chapterOutlineIsShown` doesn't depends on windows whereas `toggleChapterOutline` still depends on window dimensions, si that expected?

seanlip commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.ts:
```diff
@@ -466,7 +466,7 @@ export class StoryNodeEditorComponent implements OnInit, OnDestroy {
   toggleChapterOutline(): void {
     if (this.windowDimensionsService.isWindowNarrow()) {
-      this.chapterOutlineIsShown = !this.chapterOutlineButtonsAreShown;
+      this.chapterOutlineIsShown = !this.chapterOutlineIsShown;
```
This doesn't seem stable to me, if the window is narrow and then becomes narrow again then this will just flip its state, right? Can we make it more stable (i.e. do one thing if the window is narrow and another thing if it's not)?

seanlip commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.ts:
```diff
@@ -496,15 +496,22 @@ export class StoryNodeEditorComponent implements OnInit, OnDestroy {
-  toggleChapterOutlineButtons(): void {
-    this.chapterOutlineButtonsAreShown = !this.chapterOutlineButtonsAreShown;
+  onSaveButtonClicked(): void {
+    this.outlineEditViewIsShown = this.oldOutline;
```
Something feels wrong with this statement -- the first item is a boolean but oldOutline doesn't look like a boolean. I think some additional clarity is needed here.

seanlip reviewed (changes requested):
Thanks @hiramatsuyuusuke, took a look -- I think it's pretty close! Please also update the [PR description](https://github.com/oppia/oppia/pull/18899#issue-1903495200) with a video showing the current behaviour.

@DubeySandeep @Lawful2002 PTAL as well, thanks!

DubeySandeep reviewed (commented):
@hiramatsuyuusuke, Left one last comment, PTAL!

hiramatsuyuusuke commented:
Thanks for the reviews, @seanLip, @DubeySandeep. However it's going to take a lot of time to fix the PR. Because now I can't spend a lot of time fixing the PR. If possible, I would like someone to take over the fixes. What should I do?

seanlip commented:
Hi @hiramatsuyuusuke -- thanks for letting us know. Could you please outline whether, beyond the review comments, there's anything else needed to do the full fix? That will make it easier for someone to take over.

Thanks!

hiramatsuyuusuke commented:
Hi, @seanlip. I've done some typescript and javascript fixes, but I don't think they worked very well, so I think it's better to revert to the original and then fix it again. Then should I close this PR?

Lawful2002 commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.html:
```diff
@@ -366,6 +381,30 @@
+    width: 400px;
+  }
+  .outline-content-container:hover {
+    background: rgb(240,240,240);
```
Please use hex color codes instead of RGB.

Lawful2002 commented on core/templates/pages/story-editor-page/editor-tab/story-node-editor.component.ts:
```diff
@@ -77,6 +76,7 @@ export class StoryNodeEditorComponent implements OnInit, OnDestroy {
   editableThumbnailBgColor: string;
   oldOutline: string;
   editableOutline: string;
+  outlineEditViewIsShown: string;
```
Should this be a boolean?

Lawful2002 reviewed (changes requested):
Thanks @hiramatsuyuusuke, some nits, PTAL.

hiramatsuyuusuke committed:
Avoid hardcoding the `expect` block

hiramatsuyuusuke commented:
I'm going to move this PR forward, but I think it will slow down. And I have already unassigned me from #18366 issue. If there are any problems with proceeding with this PR, please point them out. I'd like to make it a priority for someone to take over #18366 issue.

hiramatsuyuusuke committed:
fix `this.chapterOutlineIsShown = false;`

DubeySandeep commented:
@hiramatsuyuusuke I'll take a pass/work on this PR tomorrow.
